Huawei LTE SSDP improvements

scop opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Various fixes and improvements to Huawei LTE SSDP discovery, see individual commits for details.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

scop committed:
Recognize more SSDP friendly names as "ours"

scop committed:
Fix Huawei manufacturer match

It's not a substring, but a whole string match.

scop committed:
Trigger on SoyeaLink devices

scop committed:
Pass discovered URL through context, it does not persist via user_input

scop committed:
Use manufacturer from SSDP

rytilahti reviewed (commented):
Looks straightforward to me, I'm just wondering how often a device with IGD:1 from one of those two manufacturers is not a wanted device? I.e., how useful is it to perform extra checks considering this introduces a new whitelist to maintain.

edit: I'm wondering if the routers do expose some more specific service that could be used for detection?

frenck reviewed (approved):
Good question above ☝️ 

Otherwise it looks good 👍

scop commented:
Good questions indeed.

Re the wantedness/allowlist: To be honest, I have no clue, and I have no access to devices we would not want. There used to be the huawei_router integration that dealt with some other than mobile/LTE devices, so I suppose we might encounter some. Anyway, I would be fine with just starting from a clean slate and dropping the current list, and seeing if we get any reports. (Aside, it's not a new whitelist -- one existed already, but was simpler.)  I'm kind of indifferent here, if any of you is leaning towards keeping/removing, let's do that, just let me know.

Re the services: the semi-recent device I can access at the moment exposes LANDevice:1, WANPPPConnection:1, WANConnectionDevice:1, WANDevice:1, and InternetGatewayDevice:1. None of those sounds any better or specific than the IGD we currently use, but if anyone knows better or has different opinions, do speak up.

scop commented:
Another older dongle has services WANIPConnection:1, WANConnectionDevice:1, WANCommonInterfaceConfig:1, WANDevice:1, Layer3Forwarding:1, InternetGatewayDevice:1. Its friendly name is not something that we would have found as "ours" before or after the changes in 4bd2e288c8c667e776e541b6bf6398454dcdd19a (it's just "Huawei Internet Gateway Device"). Dropping the additional match would make that one discoverable via SSDP.

Guess we could replace the friendlyname match with a probe at an endpoint that according to our best guess is known to be available without authentication in ~all devices if we're worried about false positives. If you have a better feeling about this approach than just removing the match, let me know and I'll have a look at that.

scop committed:
Match on plain "Huawei" manufacturer still

Some older supported dongles have that.

rytilahti reviewed (approved):
I think it is good as it is. After all, any potential false positives are a non-issue as the user can likely judge better and in the worst case just ignore the discovered device to stop showing it.

edit: re-reading your second comment, maybe "Huawei" should be added to the regexp to let that another dongle be discovered, too?

scop commented:
> maybe "Huawei" should be added to the regexp to let that another dongle be discovered, too?

To me the regexp exists to filter out other Huawei(/SoyeaLink) devices besides the LTE'y ones, and adding "Huawei" there would more or less defeat the purpose. In that sense I would not add it, but then again, as said, I would be open to removing the regexp match altogether and see where that takes us. As you noted, the worst case scenario wouldn't be that much of a biggie, the user has the option to ignore false positives.

scop committed:
Actually use serial number if available, update URL for configured

rytilahti commented:
Ahh, gotcha. Your call, both options are fine in my opinion.

scop committed:
Remove heuristic friendly name filtering against non-LTE devices

This is fragile at best, and actually prevents some devices from being
considered. Leave it to the user to decide instead.

scop commented:
I'm leaning towards removing it, done so in f12e37828ac387560b8ab7bba102001896828672. Going to merge this ~tomorrow unless there are objections.

rytilahti reviewed (approved):
I find it also better so, and we can always reconsider in case this will cause too many false positives. Thanks @scop :+1:

MartinHjelmare commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -248,18 +244,22 @@ async def async_step_ssdp(self, discovery_info: ssdp.SsdpServiceInfo) -> FlowRes
+                CONF_URL: url,
+            }
+        )
+        return await self._async_show_user_form()
```
Side note: We should not move directly to another step's form. The form is private to the step method. Await the step method to move to that step.
